src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject_Options.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal abstract partial class AbstractProject
+    {
+        private string _lastParsedCompilerOptions;
+        private CommandLineArguments _lastParsedCommandLineArguments;
+
+        /// <summary>
+        /// Returns the parsed command line arguments (parsed by <see cref="ParseCommandLineArguments"/>) that were set by the project
+        /// system's call to <see cref="ICompilerOptionsHostObject.SetCompilerOptions(string, out bool)"/>.
+        /// </summary>
+        protected CommandLineArguments GetParsedCommandLineArguments()
+        {
+            if (_lastParsedCommandLineArguments == null)
+            {
+                // We don't have any yet, so let's parse nothing
+                _lastParsedCompilerOptions = string.Empty;
+                _lastParsedCommandLineArguments = ParseCommandLineArguments(SpecializedCollections.EmptyEnumerable<string>());
+            }
+
+            return _lastParsedCommandLineArguments;
+        }
+
+        protected abstract CommandLineArguments ParseCommandLineArguments(IEnumerable<string> arguments);
+
+        protected void SetCommandLineArguments(string commandLine)
+        {
+            if (string.Equals(_lastParsedCompilerOptions, commandLine, StringComparison.OrdinalIgnoreCase))
+            {
+                return;
+            }
+
+            var splitArguments = CommandLineParser.SplitCommandLineIntoArguments(commandLine, removeHashComments: false);
+            var commandLineArguments = ParseCommandLineArguments(splitArguments);
+            _lastParsedCompilerOptions = commandLine;
+
+            SetCommandLineArguments(commandLineArguments);
+        }
+
+        protected void SetCommandLineArguments(CommandLineArguments commandLineArguments)
+        {
+            _lastParsedCommandLineArguments = commandLineArguments;
```

Comment:
I'll clean this up once @davkean finalizes if we need to pass in command line as a string (_lastParsedCompilerOptions would stay in AbstractProject) or as CommandLineArguments (_lastParsedCompilerOptions and ParseCommandLineArguments will move into AbstractLegacyProject)